oslabs-beta/DockerLocal
# Clone repository from Github using SSH - Linux/Mac

## Description:
- These changes implement the feature to clone a user's repository from Github. First, we create a private/public SSH key pair and add it to the user's ssh-agent. Next, we add the public SSH key to Github on behalf of the authenticated user. Then, we open an SSH connection to clone the git repository. After all repositories have been cloned, we delete the public/private SSH key files, remove the private key from the SSH agent and delete the public key from the user's Github account. 

## Changes I Made:

- [X] In .gitignore, add folders that will store clone repositories and temporary folders for SSH keys/SSH agent info.
- [X] Create sshKeyController.ts. This file contains three middleware functions: createSSHkey, addSSHkeyToGithub, and deleteSSHkey. These functions are used to handle all parts of the SSH key workflow.
- [X] Create sshKeygen.sh. This script starts an SSH agent and saves the process id. Finally, it adds the private SSH key to the SSH agent.
- [X] Create sshKeyDelete.sh. This script kills the SSH agent, removes the public key from Github and deletes the public/private key files. This is our cleanup script to make sure there are no lingering SSH keys. 
- [X] Create gitController.ts. This file creates one middleware function: cloneRepo. It will also contain the middleware for automatically pulling down changes to repositories once this functionality is added. 
- [X] Create cloneRepo.sh. This script connects to the SSH agent, makes the myRepos folder, and clones a repository over SSH from Github.
- [X] In api-route.ts, add import statements for sshKeyController and gitControlller. On the GET /api/repos route, add createSSHkey, addSSHkeyToGithub, cloneRepo, and deleteSSHkey middleware functions. 
- [X] In auth-route.ts, add "admin:public_key" scope to allow application to create/delete SSH keys on Github on behalf of the user.
- [X] In server.ts, change global error handler to log error message for debugging. This will need to be removed once application is ready for launch.

## How to Test:
- Launch application using npm start. Click Sign in With Github. When the page redirects to /api/repos, you will see a list of all the repositories associated with your account. In the current configuration, the middleware will try to clone the second repository in the list and assume you are the owner. If you receive an error indicating that the repository doesn't exist on Github, change line 33 of gitController.ts to be the index of a repo where you are the owner (E.g.   `const repoName = repos[1].name; to   const repoName = repos[2].name; `) After the repository is cloned, check to make sure there is no /tmpKeys folder in your application. In your Github settings, make sure there is no SSH key titled username@DockerLocal in your SSH keys (screenshot 2). 

![sshKeyGithub](https://user-images.githubusercontent.com/7851187/86159347-085c2d80-babf-11ea-812a-6e3c6bd1efb1.png)
![myRepos](https://user-images.githubusercontent.com/7851187/86159430-2fb2fa80-babf-11ea-920e-8c8b9e461cbb.png)
